rust/src/csvreader.rs

```diff
@@ -0,0 +1,263 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! CSV Reader
+//!
+//! This CSV reader allows CSV files to be read into the Arrow memory model. Records are loaded in
+//! batches and are then converted from row-based data to columnar data.
+//!
+//! Example:
+//!
+//! ```
+//! use arrow::csvreader::CsvReader;
+//! use arrow::datatypes::{DataType, Field, Schema};
+//! use std::fs::File;
+//! use std::sync::Arc;
+//!
+//! let schema = Schema::new(vec![
+//!   Field::new("city", DataType::Utf8, false),
+//!   Field::new("lat", DataType::Float64, false),
+//!   Field::new("lng", DataType::Float64, false),
+//! ]);
+//!
+//! let file = File::open("test/data/uk_cities.csv").unwrap();
+//!
+//! let mut csv = CsvReader::new(file, Arc::new(schema), false, 1024, None);
+//! let batch = csv.next().unwrap().unwrap();
+//!```
+
+use std::fs::File;
+use std::io::BufReader;
+use std::sync::Arc;
+
+use array::{ArrayRef, BinaryArray};
+use builder::{ArrayBuilder, ListArrayBuilder, PrimitiveArrayBuilder};
+use datatypes::{DataType, Schema};
+use error::ArrowError;
+use record_batch::RecordBatch;
+
+use csv;
+use csv::{StringRecord, StringRecordsIntoIter};
+
+/// CSV file reader
+pub struct CsvReader {
+    /// Explicit schema for the CSV file
+    schema: Arc<Schema>,
+    /// Optional projection for which columns to load (zero-based column indices)
+    projection: Option<Vec<usize>>,
+    /// File reader
+    record_iter: StringRecordsIntoIter<BufReader<File>>,
+    /// Batch size (number of records to load each time)
+    batch_size: usize,
+}
+
+impl CsvReader {
+    /// Create a new CsvReader
+    pub fn new(
+        file: File,
+        schema: Arc<Schema>,
+        has_headers: bool,
+        batch_size: usize,
+        projection: Option<Vec<usize>>,
+    ) -> Self {
+        let csv_reader = csv::ReaderBuilder::new()
+            .has_headers(has_headers)
+            .from_reader(BufReader::new(file));
+
+        let record_iter = csv_reader.into_records();
+        CsvReader {
+            schema: schema.clone(),
+            projection,
+            record_iter,
+            batch_size,
+        }
+    }
+}
+
+macro_rules! build_primitive_array {
+    ($ROWS:expr, $COL_INDEX:expr, $TY:ty) => {{
+        let mut builder = PrimitiveArrayBuilder::<$TY>::new($ROWS.len() as i64);
+        for row_index in 0..$ROWS.len() {
+            match $ROWS[row_index].get(*$COL_INDEX) {
+                Some(s) if s.len() > 0 => builder.push(s.parse::<$TY>().unwrap()).unwrap(),
+                _ => builder.push_null().unwrap(),
+            }
+        }
+        Ok(Arc::new(builder.finish()) as ArrayRef)
+    }};
+}
+
+impl CsvReader {
+    /// Read the next batch of rows
+    pub fn next(&mut self) -> Option<Result<Arc<RecordBatch>, ArrowError>> {
```

Comment:
Two points:
1. Can we return `RecordBatch` instead of `Arc<RecordBatch>`? we are creating unique reference for the record batches here and can let the caller to decide whether to make it a `Arc`.
2. We should use `error.Result` instead of `Result<T, ArrowError>`.

Reply:
done